File: units.nim
# Implementation of seconds, used for time differences.
import time_nanos, time_stamp, sequtils

proc test_blog_examples() =
  echo "Showing blog examples.\n"

  let a = "2012.01.01".date
  echo "Example 1: ", a
  echo "Example 2:"
  echo "\t", a + 1.d
  echo "\t", "2012.01.01".date + 1.d
  echo "Example 3: ", "2012.01.01".date + 1.m + 1.d + 1.h + 15.i + 17.s

  let
    r = to_seq(0 .. <10)
    offsets = r.map_it(Nano, (1.m + 1.d + 1.h + 15.i + 17.s) * it)
    values = offsets.map_it(Stamp, "2012.01.01".date + it)

  echo "Example 4: ", values

  echo "…again but compressed… ", to_seq(0 .. <10)
      .map_it(Nano, (1.m + 1.d + 1.h + 15.i + 17.s) * it)
      .map_it(Stamp, "2012.01.01".date + it)

  echo "…using helper procs… ",
    "2012.01.01".date + (1.m + 1.d + 1.h + 15.i + 17.s) * (0 .. <10)

  echo "Example 5 b[week]: ", values.map_it(int, it.week)
  echo "Example 5 b[second]: ", values.map_it(int, it.second)

  echo "\nDid all examples."

proc main() =
  test_seconds()
  test_stamps()
  test_blog_examples()

when isMainModule: main()

File: java/Units.java
import java.util.*;
import java.util.stream.*;
import myqual.Nano;
import myqual.Plain;
import myqual.Stamp;
import static abomination.NanoUtils.*;
import static abomination.StampUtils.*;
import static extra.Out.*;

public class Units
{
	public static void main(String[] args) {
		testSeconds();
		testStamps();
		testBlogExamples();
	}

	static void testBlogExamples() {
		echo("Showing blog examples.\n");

		@Stamp long a = Date("2012.01.01");
		echo("Example 1: " + sStr(a));
		echo("Example 2:");
		echo("\t" + sStr(sadd(a, d(1))));
		echo("\t" + sStr(sadd(Date("2012.01.01"), d(1))));
		echo("Example 3: " + sStr(sadd(Date("2012.01.01"),
			madd(m(1), d(1), h(1), i(15), s(17)))));

		// Nice try, java, nice try.
		List<Long> values = IntStream.range(0, 10)
			.boxed() // Hint, if the language requires boxing, it sucks.
			// First create Longs contaning the nano offsets.
			.map(intStep -> nbox(mul(
				madd(m(1), d(1), h(1), i(15), s(17)),
				(int)intStep)))
			// Apply the nano offsets to a stamp.
			.map(longNano -> sbox(sadd(Date("2012.01.01"), Nano(longNano))))
			.collect(Collectors.toList());

		echo("Example 4: @[" + values.stream()
			.map(longStamp -> sStr(Stamp(longStamp)))
			.collect(Collectors.joining(", ")) + "]");

		echo("Example 5: b[week]: @[" + values.stream()
			.map(longStamp -> String.valueOf(sWeek(Stamp(longStamp))))
			.collect(Collectors.joining(", ")) + "]");

		echo("Example 5: b[second]: @[" + values.stream()
			.map(longStamp -> String.valueOf(sSecond(Stamp(longStamp))))
			.collect(Collectors.joining(", ")) + "]");

		echo("\nDid most examples, awkwardly.");
	}
}
